File: javascript/mergeSort.js
let array = [];
let n = Math.floor(Math.random() * 1000000);
for (let i = 0; i < n; i++) {
    array[i] = Math.floor(Math.random() * 1000000000);
}

console.log("Initial Array: " + array);

let startTime = performance.now();

let p = 0;
let r = array.length - 1;

function mergeSort(A, p, r) {
    if (p < r) {
        let q = Math.floor((p + r) / 2);
        mergeSort(A, p, q);
        mergeSort(A, q + 1, r);
        merge(A, p, q, r);
    }
}

function merge(A, p, q, r) {
    let nl = q - p + 1;
    let nr = r - q;
    let L = new Array(nl);
    let R = new Array(nr);

    for (let i = 0; i < nl; i++) {
        L[i] = A[p + i];
    }
    for (let j = 0; j < nr; j++) {
        R[j] = A[q + j + 1];
    }

    let i = 0, j = 0, k = p;

    while (i < nl && j < nr) {
        if (L[i] <= R[j]) {
            A[k] = L[i];
            i++;
        } else {
            A[k] = R[j];
            j++;
        }
        k++;
    }

    while (i < nl) {
        A[k] = L[i];
        i++;
        k++;
    }

    while (j < nr) {
        A[k] = R[j];
        j++;
        k++;
    }
}

if (array.length > 0) {
    mergeSort(array, p, r);
}

let endTime = performance.now();

console.log("Sorted Array: " + array);
console.log("Execution Time: " + (endTime - startTime) + " ms!");

File: python/mergeSort.py

import time
import random
import math

A = [random.randint(1, 100) for _ in range(10)]

def merge(A, p, q, r):
	nL = q - p + 1
	nR = r - q
	L = [0] * nL
	R = [0] * nR
	for i in range(nL):
		L[i] = A[p + i]
	for j in range(nR):
		R[j] = A[q + j + 1]
	i = 0
	j = 0
	k = p
	while i < nL and j < nR:
		if L[i] < R[j]:
			A[k] = L[i]
			i = i + 1
		else:
			A[k] = R[j]
			j = j + 1
		k = k + 1
	while i < nL:
		A[k] = L[i]
		i = i + 1
		k = k + 1
	while j < nR:
		A[k] = R[j]
		j = j + 1

def mergeSort(A, p, r):
	if p < r:
		q = (p + r) // 2
		mergeSort(A, p, q)
		mergeSort(A, q + 1, r)
		merge(A, p, q, r)


A = [random.randint(1, 10) for _ in range(10)]

print("Unsorted array:", A)

startTime = time.time()

mergeSort(A, 0, len(A) - 1)

endTime = time.time()
executionTime = (endTime - startTime)

print("Sorted array:", A)
print(f"Execution time: {executionTime:.6f} seconds")
